In [3]:
import pandas as pd

In [9]:
df = pd.read_csv('../data/pokemon_data.csv')
df.head(10)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False


In [5]:
df[df['HP'] > df['Attack']].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
13,10,Caterpie,Bug,NaN,45,30,35,20,20,45,1,False
14,11,Metapod,Bug,NaN,50,20,55,25,25,30,1,False
15,12,Butterfree,Bug,Flying,60,45,50,90,80,70,1,False
16,13,Weedle,Bug,Poison,40,35,30,20,20,50,1,False
17,14,Kakuna,Bug,Poison,45,25,50,25,25,35,1,False


In [6]:
## filter only with one type( no type 2)

df[df['Type 2'].isnull()].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
9,7,Squirtle,Water,NaN,44,48,65,50,64,43,1,False
10,8,Wartortle,Water,NaN,59,63,80,65,80,58,1,False
11,9,Blastoise,Water,NaN,79,83,100,85,105,78,1,False


# with and without loc

Your two lines

## A) Boolean filtering (most common)

df[df['Type 1'] == 'Grass'].head(5)

## B) .loc version

df.loc[df['Type 1'] == 'Grass'].head(5)

more explicit pandas method

label-based indexer

allows row + column selection together



## Key answer first

    In this case, both return the same result

    BUT the way they work internally is different.

| Feature                   | `df[...]` | `df.loc[...]`  |
| ------------------------- | --------- | -------------- |
| readability               | simpler   | clearer intent |
| column selection          | limited   | powerful       |
| row + column filtering    | no        | yes            |
| recommended in interviews | yes       | YES (better)   |


# Important insight
This works:
```python
df[df['Type 1'] == 'Grass']
```
But this is more powerful:
```python
df.loc[df['Type 1'] == 'Grass', ['Name', 'HP']]
```
## .loc = full control

Why .loc is preferred in serious code

Because it avoids ambiguity:

df[...] → could be row or column confusion

df.loc → explicitly row + column selector

======================================================


## One important subtle difference

If you later do assignment:
```python
df.loc[df['Type 1'] == 'Grass', 'HP'] = 0
```
It sets the HP of all Grass-type Pokémon to 0

✔ modifies safely

But:
```python
df[df['Type 1'] == 'Grass']['HP'] = 0
```
may fail (chained indexing problem)

# Final rule

.loc is safer and more powerful

df[...] is shorter but limited

===========================================================
# To change multiple columns in pandas, you just pass a list of column names instead of a single column.

## Basic idea
```python
df.loc[rows, [col1, col2, col3]] = value
```
Example 1: change multiple columns to same value

```python
df.loc[df['Type 1'] == 'Grass', ['HP', 'Attack']] = 0
```
sets HP and Attack to 0 for Grass Pokémon

In [7]:
df.loc[df['Type 1'] ==  'Grass'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False


In [22]:
df[df['Type 1'] == 'Grass'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,1,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,2,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,3,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,3,VenusaurMega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
48,43,Oddish,Grass,Poison,45,50,55,75,65,30,1,False


In [16]:
df.loc[:,'Type 1']

0        Grass
1        Grass
2        Grass
3        Grass
4         Fire
        ...   
795       Rock
796       Rock
797    Psychic
798    Psychic
799       Fire
Name: Type 1, Length: 800, dtype: object

In [25]:
df.loc[df['Type 1'] == 'Grass', 'HP']

0       45
1       60
2       80
3       80
48      45
      ... 
718     56
719     61
720     88
740     66
741    123
Name: HP, Length: 70, dtype: int64

In [26]:
df.loc[(df['Type 1'] == 'Grass') & (df['HP'] > 100), ['Name', 'HP']]

,Name,HP
652,Amoonguss,114
741,Gogoat,123


In [33]:
df.loc[df['Type 1'] == 'Fire'].head(5)

,#,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
4,4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
5,5,Charmeleon,Fire,NaN,58,64,58,80,65,80,1,False
6,6,Charizard,Fire,Flying,78,84,78,109,85,100,1,False
7,6,CharizardMega Charizard X,Fire,Dragon,78,130,111,130,85,100,1,False
8,6,CharizardMega Charizard Y,Fire,Flying,78,104,78,159,115,100,1,False


In [ ]:
df.loc[df['HP'] > 100, ['Name', 'HP'] ].head(10)

,Name,HP
44,Jigglypuff,115
45,Wigglytuff,140
96,Muk,105
120,Rhydon,105
121,Chansey,250
123,Kangaskhan,105
124,KangaskhanMega Kangaskhan,105
142,Lapras,130
145,Vaporeon,130
155,Snorlax,160


Here are some pandas filtering challenges using a Pokémon dataset (assuming columns like `Name`, `Type 1`, `Type 2`, `HP`, `Attack`, `Defense`, `Speed`, `Legendary`, `Generation`).

Try writing the `.loc[]` filters yourself before checking the answers.

### Easy

1. Get all Grass Pokémon.
2. Get all Pokémon with HP greater than 100.
3. Get all Legendary Pokémon.
4. Get all Pokémon from Generation 1.
5. Get only the `Name` and `Attack` columns for Fire Pokémon.

```python
# Example format
df.loc[df['Type 1'] == 'Grass']
```

---

### Medium

6. Get all Water Pokémon with HP above 80.
7. Get all Pokémon whose Attack is greater than Defense.
8. Get all Pokémon that are either Fire or Electric.
9. Get all non-Legendary Pokémon.
10. Get Pokémon with Speed between 80 and 120.

---

### Hard

11. Get Pokémon that are Grass AND Poison.
12. Get Pokémon with HP > 100 and Attack > 100.
13. Get Pokémon that are Legendary OR have Attack > 150.
14. Get Pokémon that are not Water type.
15. Get Pokémon whose name starts with `"Char"`.

Hint:

```python
df['Name'].str.startswith('Char')
```

---

### Very Hard

16. Get Pokémon whose Type 1 is Fire but Type 2 is NOT Flying.
17. Get Pokémon with Attack in the top 10% of the dataset.
18. Get all Pokémon where HP is above the dataset average.
19. Get Pokémon that have either Grass or Water in **Type 1 or Type 2**.
20. Get all Pokémon whose name contains `"saur"` (case insensitive).

Hint:

```python
df['Name'].str.contains('saur', case=False)
```

---

### Coordinates Practice (`loc[row_filter, column_filter]`)

21. Return only the names of Grass Pokémon.

22. Return the HP and Attack of Legendary Pokémon.

23. Return the Name, HP, and Speed of Pokémon with HP > 120.

24. Return only the Attack column for Fire Pokémon.

25. Return all columns for Pokémon named "Pikachu".

---

### Challenge Mode

Without using Google:

26. Fire Pokémon with Attack > 100 and Speed > 100.

27. Non-Legendary Pokémon from Generation 1 with HP > 80.

28. Pokémon whose Type 1 is Water and Type 2 is Ice.

29. Pokémon with Attack greater than the average Attack of the dataset.

30. Pokémon that satisfy:

* HP > 80
* Attack > 100
* Speed > 100
* Not Legendary
